docs/setup_leakcheck.md

```diff
@@ -0,0 +1,167 @@
+```console
+$ make main.exe
+$ leaks -quiet -atExit -- ./main.exe
+```
+
+## Prerequisites
+This tutorial relies on command line tools.  If you haven't installed CLI tools on your machine yet, follow one of these tutorials first.
+
+| [macOS](setup_macos.html)| [Windows](setup_wsl.html) | [Linux](setup_wsl.html#install-cli-tools)
+
+We're also assuming you are familiar with a text editor or IDE.  If you haven't already done so, follow one of these tutorials.
+
+| [VS Code (recommended)](https://eecs280staff.github.io/tutorials/setup_vscode.html)| [Visual Studio](https://eecs280staff.github.io/tutorials/setup_visualstudio.html) | [Xcode](https://eecs280staff.github.io/tutorials/setup_xcode.html) |
+
+Finally, you'll need to be familiar with the command line interface (CLI).  If you haven't seen it yet, take a look at the [Command Line Tutorial](cli.html).
+
+## Example program
+We'll work in a directory called `leakcheck` just for this tutorial.
+```console
+$ mkdir leakcheck
+```
+
+Create a file called `main.cpp` and copy-paste this leaky Hello World program.
+```c++
+#include <iostream>
+using namespace std;
+
+int main() {
+  new int;  // Memory leak!
+  cout << "Hello Leaks!\n";
+}
+```
+{: data-title="main.cpp" }
+
+Create a file called `Makefile` and copy-paste this code.
+```make
+CXX ?= g++
+CXXFLAGS ?= -Wall -Werror -pedantic -g --std=c++11 -Wno-sign-compare -Wno-comment
+
+# Run regression test
+test: main.exe
+	./main.exe
+
+# Compile the main executable
+main.exe: main.cpp
+	$(CXX) $(CXXFLAGS) main.cpp -o main.exe
+
+# Remove automatically generated files
+clean :
+	rm -rvf *.exe *~ *.out *.dSYM *.stackdump
+```
+{: data-title="Makefile" }
+
+Your directory should look like this:
+```console
+$ tree
+.
+├── Makefile
+└── main.cpp
+```
+
+The program should compile and run.
+```console
+$ make main.exe
+$ ./main.exe
+Hello Leaks!
+```
+
+## macOS
+Verify that your Makefile does *not* enable the address sanitizer.
+```make
+CXXFLAGS ?= -Wall -Werror -pedantic -g --std=c++11 -Wno-sign-compare -Wno-comment
+# primer-spec-highlight-start
+# Should *not* see -fsanitize=address
+# primer-spec-highlight-end
+```
+{: data-title="Makefile" }
+
+Compile and run with the `leaks` program.
+```console
+$ make main.exe
+$ leaks -quiet -atExit -- ./main.exe
+main.exe(14025) MallocStackLogging: could not tag MSL-related memory as no_footprint, so those pages will be included in process footprint - (null)
+main.exe(14025) MallocStackLogging: recording malloc and VM allocation stacks using lite mode
+Hello Leaks!
+Process 14025 is not debuggable. Due to security restrictions, leaks can only show or save contents of readonly memory of restricted processes.
+
+leaks Report Version: 4.0, multi-line stacks
+Process 14025: 226 nodes malloced for 17 KB
+Process 14025: 1 leak for 16 total leaked bytes.
+
+STACK OF 1 INSTANCE OF 'ROOT LEAK: <malloc in main>':
```

Comment:
Did you disable the address sanitizer?  Are you on x86_64 or arm64?

Reply:
> Did you disable the address sanitizer?

Yes, by using the exact Makefile given in the tutorial (no -fsanitize=address)

> Are you on x86_64 or arm64?

I am on arm64 (2021 MacBook Pro, M1 Pro)

Reply:
It looks like the difference might be whether `MallocStackLogging` is enabled? I needed to do this to get line numbers.

@ohjuny Could you try running `export MallocStackLogging=1` and then rerun `leaks`?

Reply:
Actually, don't do that. It's obnoxious to `export MallocStackLogging=1` since then many commands will start producing a bunch of annoying output.

Instead, try:
```console
MallocStackLogging=1 leaks -quiet -atExit -- ./main.exe
```

Reply:
I get nearly identical output with and without `MallocStackLogging=1` on my arm64 machine.  I did find that if I omit the `-g` compiler flag I don't see line numbers (no surprise).

Reply:
For completeness, here's what I see:

```console
jjuett@Jamess-MacBook-Pro published % make clean   
rm -vrf *.o *.exe *.gch *.dSYM *.stackdump *.out
jjuett@Jamess-MacBook-Pro published % c++ --version
Apple clang version 14.0.0 (clang-1400.0.29.102)
Target: arm64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
jjuett@Jamess-MacBook-Pro published % make List_compile_check.exe
c++ --std=c++11 -Wall -Werror -pedantic -g -Wno-sign-compare -Wno-comment List_compile_check.cpp -o List_compile_check.exe
jjuett@Jamess-MacBook-Pro published % leaks -quiet -atExit -- ./List_compile_check.exe       
1
0
42
42
42
3
1
Process 32330 is not debuggable. Due to security restrictions, leaks can only show or save contents of readonly memory of restricted processes.
 
leaks Report Version: 4.0
Process 32330: 218 nodes malloced for 16 KB
Process 32330: 4 leaks for 128 total leaked bytes.
 
    4 (128 bytes) << TOTAL >>
 
      2 (64 bytes) ROOT CYCLE: 0x128606830 [32]
         1 (32 bytes) ROOT CYCLE: 0x128606850 [32]
            CYCLE BACK TO 0x128606830 [32]
 
      1 (32 bytes) ROOT LEAK: 0x128606870 [32]
      1 (32 bytes) ROOT LEAK: 0x128606890 [32]
 
jjuett@Jamess-MacBook-Pro published % MallocStackLogging=1 leaks -quiet -atExit -- ./List_compile_check.exe      
leaks(32341) MallocStackLogging: could not tag MSL-related memory as no_footprint, so those pages will be included in process footprint - (null)
leaks(32341) MallocStackLogging: stack logs being written to /tmp/stack-logs.32341.1041d4000.leaks.zrbji2.index
leaks(32341) MallocStackLogging: recording malloc and VM allocation stacks to disk using standard recorder
List_compile_check.exe(32342) MallocStackLogging: could not tag MSL-related memory as no_footprint, so those pages will be included in process footprint - (null)
List_compile_check.exe(32342) MallocStackLogging: stack logs being written to /tmp/stack-logs.32342.1026bc000.List_compile_check.exe.B7qe83.index
List_compile_check.exe(32342) MallocStackLogging: recording malloc and VM allocation stacks to disk using standard recorder
1
0
42
42
42
3
1
List_compile_check.exe(32342) MallocStackLogging: stack logs deleted from /tmp/stack-logs.32342.1026bc000.List_compile_check.exe.B7qe83.index
Process 32342 is not debuggable. Due to security restrictions, leaks can only show or save contents of readonly memory of restricted processes.
 
leaks Report Version: 4.0, multi-line stacks
Process 32342: 221 nodes malloced for 16 KB
Process 32342: 4 leaks for 128 total leaked bytes.
 
STACK OF 1 INSTANCE OF 'ROOT CYCLE: malloc<32>':
4   dyld                                  0x1025e508c start + 520
3   List_compile_check.exe                0x1024d6c84 main + 728  List_compile_check.cpp:48
2   List_compile_check.exe                0x1024d7250 List<int>::push_back(int const&) + 36  List.h:217
1   libc++abi.dylib                       0x1a69508a0 operator new(unsigned long) + 32
0   libsystem_malloc.dylib                0x1a67bf7bc _malloc_zone_malloc + 272 
====
    2 (64 bytes) ROOT CYCLE: 0x151606a10 [32]
       1 (32 bytes) ROOT CYCLE: 0x151606a30 [32]
          CYCLE BACK TO 0x151606a10 [32]
 
STACK OF 1 INSTANCE OF 'ROOT LEAK: malloc<32>':
4   dyld                                  0x1025e508c start + 520
3   List_compile_check.exe                0x1024d6f18 main + 1388  List_compile_check.cpp:68
2   List_compile_check.exe                0x1024d7250 List<int>::push_back(int const&) + 36  List.h:217
1   libc++abi.dylib                       0x1a69508a0 operator new(unsigned long) + 32
0   libsystem_malloc.dylib                0x1a67bf7bc _malloc_zone_malloc + 272 
====
    1 (32 bytes) ROOT LEAK: 0x151606920 [32]
 
STACK OF 1 INSTANCE OF 'ROOT LEAK: malloc<32>':
4   dyld                                  0x1025e508c start + 520
3   List_compile_check.exe                0x1024d6f8c main + 1504  List_compile_check.cpp:72
2   List_compile_check.exe                0x1024d7250 List<int>::push_back(int const&) + 36  List.h:217
1   libc++abi.dylib                       0x1a69508a0 operator new(unsigned long) + 32
0   libsystem_malloc.dylib                0x1a67bf7bc _malloc_zone_malloc + 272 
====
    1 (32 bytes) ROOT LEAK: 0x151606940 [32]
 
leaks(32341) MallocStackLogging: stack logs deleted from /tmp/stack-logs.32341.1041d4000.leaks.zrbji2.index
```